rust/connlib/libs/client/src/lib.rs

```diff
@@ -1,25 +1,246 @@
+        runtime_stopper: tokio::sync::mpsc::Sender<StopRuntime>,
+        portal_url: Url,
+        token: String,
+        device_id: String,
+        callbacks: CallbackErrorFacade<CB>,
+    ) {
+        runtime.spawn(async move {
+            let private_key = StaticSecret::random_from_rng(rand::rngs::OsRng);
+            let name_suffix: String = thread_rng().sample_iter(&Alphanumeric).take(8).map(char::from).collect();
+            let external_id = sha256(device_id);
+
+            let connect_url = fatal_error!(
+                get_websocket_path(portal_url, token, "client", &Key(PublicKey::from(&private_key).to_bytes()), &external_id, &name_suffix),
+                runtime_stopper,
+                &callbacks
+            );
+
+            // This is kinda hacky, the buffer size is 1 so that we make sure that we
+            // process one message at a time, blocking if a previous message haven't been processed
+            // to force queue ordering.
+            let (control_plane_sender, mut control_plane_receiver) = tokio::sync::mpsc::channel(1);
+
+            let mut connection = PhoenixChannel::<_, IngressMessages, ReplyMessages, Messages>::new(connect_url, move |msg, reference| {
+                let control_plane_sender = control_plane_sender.clone();
+                async move {
+                    tracing::trace!("Received message: {msg:?}");
+                    if let Err(e) = control_plane_sender.send((msg, reference)).await {
+                        tracing::warn!("Received a message after handler already closed: {e}. Probably message received during session clean up.");
+                    }
+                }
+            });
+
+            let control_signaler = ControlSignaler { control_signal: connection.sender_with_topic("client".to_owned()) };
+            let tunnel = fatal_error!(
+                Tunnel::new(private_key, control_signaler.clone(), callbacks.clone()).await,
+                runtime_stopper,
+                &callbacks
+            );
+
+            let mut control_plane = ControlPlane {
+                tunnel: Arc::new(tunnel),
+                control_signaler,
+                tunnel_init: Mutex::new(false),
+            };
+
+            tokio::spawn(async move {
+                let mut interval = tokio::time::interval(Duration::from_secs(10));
+                loop {
+                    tokio::select! {
+                        Some((msg, reference)) = control_plane_receiver.recv() => {
+                            match msg {
+                                Ok(msg) => control_plane.handle_message(msg, reference).await?,
+                                Err(err) => control_plane.handle_error(err, reference).await,
+                            }
+                        },
+                        _ = interval.tick() => control_plane.stats_event().await,
+                        else => break
+                    }
+                }
+                Result::Ok(())
+            });
```

Comment:
Now that everything is in one place here and not generic, we can - in a future PR - make use of the other `phoenix-channel` library here and write an event-loop that reads a message from the websocket channel and directly calls a handler function on the `ControlPlane` (or `Tunnel`, depending on how many interfaces we want to simplify).